npm/nameSpaceController_test.go

```diff
@@ -0,0 +1,482 @@
+// Copyright 2018 Microsoft. All rights reserved.
+// MIT License
+package npm
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/Azure/azure-container-networking/npm/ipsm"
+	"github.com/Azure/azure-container-networking/npm/util"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubeinformers "k8s.io/client-go/informers"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	core "k8s.io/client-go/testing"
+)
+
+var (
+	alwaysReady        = func() bool { return true }
+	noResyncPeriodFunc = func() time.Duration { return 0 }
+)
+
+type expectedNsValues struct {
+	expectedLenOfPodMap    int
+	expectedLenOfNsMap     int
+	expectedLenOfWorkQueue int
+}
+
+type nameSpaceFixture struct {
+	t *testing.T
+
+	kubeclient *k8sfake.Clientset
+	// Objects to put in the store.
+	nsLister []*corev1.Namespace
+	// Actions expected to happen on the client.
+	kubeactions []core.Action
+	// Objects from here preloaded into NewSimpleFake.
+	kubeobjects []runtime.Object
+
+	// (TODO) will remove npMgr if possible
+	npMgr        *NetworkPolicyManager
+	ipsMgr       *ipsm.IpsetManager
+	nsController *nameSpaceController
+	kubeInformer kubeinformers.SharedInformerFactory
+}
+
+func newNsFixture(t *testing.T) *nameSpaceFixture {
+	f := &nameSpaceFixture{
+		t:           t,
+		nsLister:    []*corev1.Namespace{},
+		kubeobjects: []runtime.Object{},
+		npMgr:       newNPMgr(t),
+		ipsMgr:      ipsm.NewIpsetManager(),
+	}
+	return f
+}
+
+func (f *nameSpaceFixture) newNsController() {
+	f.kubeclient = k8sfake.NewSimpleClientset(f.kubeobjects...)
+	f.kubeInformer = kubeinformers.NewSharedInformerFactory(f.kubeclient, noResyncPeriodFunc())
+
+	f.nsController = NewNameSpaceController(f.kubeInformer.Core().V1().Namespaces(), f.kubeclient, f.npMgr)
+	f.nsController.nameSpaceListerSynced = alwaysReady
+
+	for _, ns := range f.nsLister {
+		f.kubeInformer.Core().V1().Namespaces().Informer().GetIndexer().Add(ns)
+	}
+}
+
+func (f *nameSpaceFixture) ipSetSave(ipsetConfigFile string) {
+	//  call /sbin/ipset save -file /var/log/ipset-test.conf
+	f.t.Logf("Start storing ipset to %s", ipsetConfigFile)
+	if err := f.ipsMgr.Save(ipsetConfigFile); err != nil {
+		f.t.Errorf("TestAddPod failed @ ipsMgr.Save")
+	}
+}
+
+func (f *nameSpaceFixture) ipSetRestore(ipsetConfigFile string) {
+	//  call /sbin/ipset restore -file /var/log/ipset-test.conf
+	f.t.Logf("Start re-storing ipset to %s", ipsetConfigFile)
+	if err := f.ipsMgr.Restore(ipsetConfigFile); err != nil {
+		f.t.Errorf("TestAddPod failed @ ipsMgr.Restore")
+	}
+}
+func newNPMgr(t *testing.T) *NetworkPolicyManager {
+	npMgr := &NetworkPolicyManager{
+		NsMap:            make(map[string]*Namespace),
+		PodMap:           make(map[string]*NpmPod),
+		RawNpMap:         make(map[string]*networkingv1.NetworkPolicy),
+		ProcessedNpMap:   make(map[string]*networkingv1.NetworkPolicy),
+		TelemetryEnabled: false,
```

Comment:
For NSController UT, do we need podMap, RawNpMap, ProcessNpMap, and TelemetryEnabled"?